### DISCUSSION TITLE
boost::asio based scheduler

### DISCUSSION THREAD

**Discussion by geiseri** [victimsnino/ReactivePlusPlus#350] [Q&A] [answered]
I am trying to implement a scheduler that can operate on Asio [executors](https://www.boost.org/doc/libs/1_82_0/doc/html/boost_asio/reference/Executor1.html).   So far I have used the Qt event loop scheduler as an example. In there, you dispatch on the Qt event loop utilizing a timer to get the function to execute in the eventloop's context.  I would like to do the same, but with a specific executor and not a singleton as Qt's main event loop is.  I think the [runloop_scheduler](https://github.com/victimsnino/ReactivePlusPlus/blob/v1/src/rpp/rpp/schedulers/run_loop_scheduler.hpp) is what I want to use, but it is not clear how I want to manage the `dispatch_if_ready`/`dispatch` calls.  Is as simple as calling `create_worker` and keeping the result around to call the dispatch methods?  Can I just keep it around and pass it into `observe_on`?  If so what owns that instance?  The snake example seems to imply it must be a static instance.   Did I miss something?

**Comment by AlexInLog** [comment, marked as answer]
About runloop_scheduler: it doesn't requires that runloop_scheduler should be static or something like this. It snake example it is global just for simplicity (prevent passing it as argument to each method). In general it works like this:
1) you create `runloop_scheduler`. Scheduler owns shared_ptr to queue of schedulables
2) when you pass scheduler to some operator (like `observe_on`), then it just keeps copy of this one
3) `observer_on` calls `create_worker` during subscription for passed subscriber. At this moment we have separate and uniq worker for specific subscriber. This worker keeps weak_ptr to same queue as original scheduler
4) any later schedulings via worker (actual scheduling of emissions) would try to obtain shared_ptr from weak_ptr and pass schedulables to it.
5) to PROCEED schedulables we need manually dispatch it via scheduler in some other place (in main() in snake case)

So, it is kind of "manual" scheduler: you decide when and how many schedulables you want to proceed at the specific moment.


I'm not familiar with asio executors, but looks like you need to pass context/executor inside your new custom scheduler and pass it to required operator. Operator would call `create_worker`. At this point you need to handle context/executor somehow (create resolver or something like this???). And then schedule execution via context/executor/resolver on actual scheduling via this worker.
